lib/src/third_party/diff_match_patch/utils.dart

```diff
@@ -0,0 +1,253 @@
+  final lineArray = <String>[];
+  final lineHash = HashMap<String, int>();
+  // e.g. linearray[4] == 'Hello\n'
+  // e.g. linehash['Hello\n'] == 4
+
+  // '\x00' is a valid character, but various debuggers don't like it.
+  // So we'll insert a junk entry to avoid generating a null character.
+  lineArray.add('');
+
+  // Allocate 2/3rds of the space for text1, the rest for text2.
+  final chars1 = _diffLinesToCharsMunge(text1, lineArray, lineHash, 40000);
+  final chars2 = _diffLinesToCharsMunge(text2, lineArray, lineHash, 65535);
+
+  return {'chars1': chars1, 'chars2': chars2, 'lineArray': lineArray};
+}
+
+/// Compute the Levenshtein distance; the number of inserted, deleted or
+/// substituted characters.
+/// [diffs] is a List of Diff objects.
+/// Returns the number of changes.
+int diffLevenshtein(List<Diff> diffs) {
+  var levenshtein = 0;
+  var insertions = 0;
+  var deletions = 0;
+
+  for (var aDiff in diffs) {
+    switch (aDiff.operation) {
+      case Operation.insert:
+        insertions += aDiff.text.length;
+        break;
+      case Operation.delete:
+        deletions += aDiff.text.length;
+        break;
+      case Operation.equal:
+        // A deletion and an insertion is one substitution.
+        levenshtein += max(insertions, deletions);
+        insertions = 0;
+        deletions = 0;
+        break;
+    }
+  }
+
+  levenshtein += max(insertions, deletions);
+  return levenshtein;
+}
+
+int diffLevenshteinWord(Iterable<Diff> diffs) {
+  var levenshtein = 0;
+  var insertions = 0;
+  var deletions = 0;
+
+  for (var aDiff in diffs) {
+    switch (aDiff.operation) {
+      case Operation.insert:
+        insertions += _countWords(aDiff.text);
+        break;
+      case Operation.delete:
+        deletions += _countWords(aDiff.text);
+        break;
+      case Operation.equal:
+        // A deletion and an insertion is one substitution.
+        levenshtein += max(insertions, deletions);
+        insertions = 0;
+        deletions = 0;
+        break;
+    }
+  }
+
+  levenshtein += max(insertions, deletions);
+  return levenshtein;
+}
+
+int _countWords(String text) {
+  if (text.isEmpty) {
+    return 0;
+  }
+
+  return text.trimRight().split(' ').length;
+}
+
+@visibleForTesting
+List<String> diffRebuildtexts(List<Diff> diffs) {
+  // Construct the two texts which made up the diff originally.
+  final text1 = StringBuffer();
+  final text2 = StringBuffer();
+  for (var x = 0; x < diffs.length; x++) {
+    if (diffs[x].operation != Operation.insert) {
+      text1.write(diffs[x].text);
+    }
+    if (diffs[x].operation != Operation.delete) {
+      text2.write(diffs[x].text);
+    }
+  }
+  return [text1.toString(), text2.toString()];
+}
+
+/// Determine if the suffix of one string is the prefix of another.
+///
+/// [text1] is the first string.
+/// [text2] is the second string.
+/// Returns the number of characters common to the end of the first
+/// string and the start of the second string.
```

Comment:
This should be the first line of the dartdoc. It describes what the function actually does.